File: resize.rb
require 'optparse'
require 'fileutils'

# Option parser to call script from command line
## Options struct to contain geometry, size, and primary partition designation
Options = Struct.new(:geometry, :size, :primary, :device)

options = Options.new("#{FileUtils.pwd()}/sda-pt.sf", 1000000000, 2, "/dev/sda")
OptionParser.new do |opts|

  opts.banner = "Usage: resize.rb [options]"

  opts.on("-g", "--geometry FILEPATH", "Use specificed geometry file") do |sf|
    options.geometry = sf
  end

  opts.on("-s", "--size DISKSIZE", "Specify disk size in sectors") do |size|
    options.size = size
  end

  opts.on("-p", "--primary PARTNUM", "Specify primary partition by number") do |primary|
    options.primary = primary
  end

  opts.on("-d", "--device DEVICE", "Specify block device (ex. /dev/sda)") do |device|
    options.device = device
  end
end.parse!


geometryFile = File.readlines("#{options.geometry}")

#Starting at end of file, move back through partitions until primary partition
lineIndex = -1
partEnd = options.size.to_i
until geometryFile[lineIndex].include? "#{options.device}#{options.primary}"
  partSize = geometryFile[lineIndex].scan(/size= *\d*/)[0].split(" ")[1].to_i
  partStart = partEnd-partSize
  geometryFile[lineIndex].gsub!(/start= *\d*/, "start= #{partStart}")
  lineIndex -= 1
  partEnd = partStart
end

# generate primary partition specification and output full spec
partStart = geometryFile[lineIndex].scan(/start= *\d*/)[0].split(" ")[1].to_i
partSize = partEnd-partStart

geometryFile[lineIndex].gsub!(/size= *\d*/, "size=#{partSize}")

puts geometryFile

File: resize.py
import argparse
import os
import sys
import re

"""Argument Parser
Accepts device, geometry, and primary arguments"""
parser = argparse.ArgumentParser()
parser.add_argument('--device', help='Specify block device to create partition table for (defaults to /dev/sda)', default="/dev/sda")
parser.add_argument('--geometry', help='Specify sfpart geometry file (defaults to $PWD/sda-pt.sf)', default=f"{os.getcwd()}/sda-pt.sf")
parser.add_argument('--primary', help='Specify primary (resized) partition (defaults to 2)', default=2)
parser.add_argument('--size', help='Override size of --device')

args = parser.parse_args()
blockDevice = args.device.replace("/dev/", "")

"""Get the number of sectors on the disk"""
diskSectors = open(f"/sys/block/{blockDevice}/size").read().rstrip()
if args.size is not None:
    diskSectors = int(args.size)

with open(args.geometry) as geometryFile:
    geometry = [line.rstrip() for line in geometryFile]

def getPartitionSize(partitionLine):
    return int(re.search(r"size= *\d*", partitionLine).group().replace("size=", ""))

def getPartitionStart(partitionLine):
    return int(re.search(r"start= *\d*", partitionLine).group().replace("start=", ""))

if int(diskSectors) < (getPartitionSize(geometry[-1]) + getPartitionStart(geometry[-1])):
    sys.exit("Oops!  Looks like the disk you're trying to make this for is too small!")

lineIndex = -1

partEnd = int(diskSectors)

while True:
    geoLine = geometry[lineIndex]
    if f"{args.device}{args.primary}" in geoLine:
        break
    else:
        partStart = partEnd - getPartitionSize(geoLine)
        geometry[lineIndex] = re.sub(r"start= *\d*", f"start= {partStart}", geoLine)
        partEnd = partStart
        lineIndex -= 1

geometry[lineIndex] = re.sub(r"size= *\d*", f"size= {partEnd - getPartitionStart(geometry[lineIndex])}", geometry[lineIndex])


for line in geometry:
    print(line)

